gradle-plugin-core/src/main/java/wtf/emulator/setup/ProjectConfigurator.java

```diff
@@ -55,6 +75,87 @@ public void configure() {
+        });
+  }
+
+  @SuppressWarnings("EagerGradleConfiguration") // we want to eagerly configure added devices
+  private void registerManagedDeviceExtension() {
+    CommonExtension<?, ?, ?, ?, ?> androidCommonExtension = target.getExtensions().findByType(CommonExtension.class);
+
+    if (androidCommonExtension == null) {
+      target.getLogger().debug("Android common extension not found. Skipping ewDevices setup for project {}", target.getPath());
+      return;
+    }
+
+    ManagedDevices managedDevices = androidCommonExtension.getTestOptions().getManagedDevices();
+    ObjectFactory objects = target.getObjects();
+
+    ExtensionAware extensionAwareManagedDevices = (ExtensionAware) managedDevices;
+    ExtraPropertiesExtension extraProperties = extensionAwareManagedDevices.getExtensions().getExtraProperties();
+
+    if (extraProperties.has(EW_DEVICES_CONTAINER_KEY)) {
+      return;
+    }
+
+    EwManagedDeviceFactory ewDeviceFactory = objects.newInstance(EwManagedDeviceFactory.class);
+
+    NamedDomainObjectContainer<EwManagedDevice> ewDevicesContainer =
+      objects.domainObjectContainer(EwManagedDevice.class, ewDeviceFactory);
+    extraProperties.set(EW_DEVICES_CONTAINER_KEY, ewDevicesContainer);
+
+    // When an EwManagedDevice is added to our ewDevices container, add it to allDevices
+    ewDevicesContainer.whenObjectAdded(ewDevice -> {
+      if (managedDevices.getAllDevices().findByName(ewDevice.getName()) == null ||
+        !(managedDevices.getAllDevices().findByName(ewDevice.getName()) instanceof EwManagedDevice)) {
+        managedDevices.getAllDevices().add(ewDevice);
+      }
+    });
+    // When an EwManagedDevice is removed from our ewDevices container, remove it from allDevices
+    ewDevicesContainer.whenObjectRemoved(ewDevice -> {
+      Device deviceInAll = managedDevices.getAllDevices().findByName(ewDevice.getName());
+      if (deviceInAll instanceof EwManagedDevice) { // Make sure it's the correct type
+        managedDevices.getAllDevices().remove(deviceInAll);
+      }
+    });
+
+    // When any Device is added to allDevices, if it's an EwManagedDevice, add it to our ewDevices container
+    managedDevices.getAllDevices().whenObjectAdded(device -> {
+      if (device instanceof EwManagedDevice) {
+        EwManagedDevice ewDevice = (EwManagedDevice) device;
+        if (ewDevicesContainer.findByName(ewDevice.getName()) == null) {
+          ewDevicesContainer.add(ewDevice);
+        }
+      }
+    });
+    // When any Device is removed from allDevices, if it's an EwManagedDevice, remove it from our ewDevices container
+    managedDevices.getAllDevices().whenObjectRemoved(device -> {
+      if (device instanceof EwManagedDevice) {
+        EwManagedDevice ewDevice = (EwManagedDevice) device;
+        EwManagedDevice deviceInEw = ewDevicesContainer.findByName(ewDevice.getName());
+        if (deviceInEw != null) {
+          ewDevicesContainer.remove(deviceInEw);
+        }
+      }
+    });
```

Comment:
do we replicate both ways just for symmetry or is `ewDevices` container actually used somewhere?